Update nginx/ingress controller to 0.9.0 release

cap10morgan opened the pull request:
*****************************************************************************************

Thank you for contributing to kubernetes/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/kubernetes/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/kubernetes/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/kubernetes/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the 
history. This will make it easier to identify new changes. The PR will be squashed 
anyways when it is merged. Thanks.

*****************************************************************************************

cap10morgan committed:
Update nginx/ingress controller to 0.9.0 release

cap10morgan committed:
Update config defaults in nginx-ingress README

cap10morgan commented:
/assign @mattfarina

mattfarina commented on stable/nginx-ingress/Chart.yaml:
```diff
@@ -1,6 +1,6 @@
 name: nginx-ingress
 version: 0.8.22
```
The chart version needs to be incremented. This is causing a CI failure at the moment.

mattfarina commented on stable/nginx-ingress/values.yaml:
```diff
@@ -4,8 +4,8 @@
   image:
-    repository: k8s.gcr.io/nginx-ingress-controller
-    tag: "0.9.0-beta.15"
+    repository: quay.io/kubernetes-ingress-controller/nginx-ingress-controller
```
Note, for anyone else who comes along... the upstream project moved to quay.

mattfarina commented on stable/nginx-ingress/templates/controller-deployment.yaml:
```diff
@@ -40,6 +40,9 @@ spec:
           args:
             - /nginx-ingress-controller
             - --default-backend-service={{ if .Values.defaultBackend.enabled }}{{ .Release.Namespace }}/{{ template "nginx-ingress.defaultBackend.fullname" . }}{{ else }}{{ .Values.controller.defaultBackendService }}{{ end }}
+          {{- if (eq "0.9.0" .Values.controller.image.tag) }}
```
This will only work for 0.9.0. 0.9.1, for example, won't have the annotation prefix applied. Did you intent for it to be a range of versions? Version comparison functions are available... https://masterminds.github.io/sprig/semver.html

cap10morgan committed:
Bump nginx-ingress chart version to 0.8.23

cap10morgan committed:
Use semverCompare for nginx-ingress annotation prefix

cap10morgan committed:
Add nginx-ingress annotation prefix to daemonset

cap10morgan commented:
Thanks for the review, @mattfarina. I believe I've addressed your comments and I added the annotation prefix change to the daemonset controller too. 0.9.0 is changing the default annotation prefix, but keeping it compatible with existing installs seemed like a good thing. But it also seems like this chart tends to use the ingress class annotation instead (which is not changing FWICT)? Would you rather I left the annotation prefix arg out altogether so that new chart installs get the new 0.9.0 default?

mattfarina commented:
@cap10morgan We want compatibility with N and N-1 versions of Kubernetes. Eventually we want to get to N-2 versions. So, 1.9 and 1.8 with 1.7 as a nice to have. If the generated configuration is different for different versions that's good to do. There are other places in the charts repo already doing that.

If you can keep compatibility with current installs that's nice to do. It's not a hard requirement as there are lots of reasons to break compat (but you can't make breaking changes and only bump a patch version).

mattfarina commented:
@cap10morgan in the release 0.9-beta.18 the annotations changed. This is a breaking change. The minor version should be incremented for the version of this chart.

mattfarina commented:
/ok-to-test

cap10morgan commented:
@mattfarina W/r/t your breaking change / minor version bump comment, this is the breaking change that my `--annotations-prefix` addition addresses (and should render non-breaking). Do you want me to leave that in and leave this with just a patch version bump or take it out, bump the minor version, and note the breaking change in the changelog somewhere? (Or something else entirely?)

mattfarina commented:
@cap10morgan Should the use of `--annotations-prefix` to keep the old behavior be behind a property? While I understand keeping compatibility, should that dictate all of the future?

unguiculus commented:
I would not set `--annotations-prefix`. Neither would I put it behind a property. It is already possible to set it with current functionality. You may add a note to the readme. I'm already using the chart with this image and without this flag. I would expect people to read release notes and not blindly update stuff. Let's make this a `0.9.0` instead of a `0.8.23`.

unguiculus commented:
/hold

orangejulius commented:
Hi!
I would just like to drop by and say that upgrading the nginx ingress from the current chart version (0.9.0-beta.15) to 0.9.0 fixed an issue in my deployment. I'm not exactly sure of the cause, but it was appearing as a 3-5 minute downtime every 15 minutes for a service hosted through the ingress. The service when accessed directly from within the cluster always worked. So hopefully others will benefit from this PR being merged :)

cap10morgan commented:
Closing in favor of #3462.
